File: qr-scanner.m
#import <Cocoa/Cocoa.h>
#import <AVFoundation/AVFoundation.h>
#import <CoreImage/CoreImage.h>
#import <CoreMedia/CoreMedia.h>
#import <CoreVideo/CoreVideo.h>

@interface ScannerOverlayView : NSView
@property (nonatomic, assign) CGFloat progress;
@property (nonatomic, retain) NSString *statusText;
@property (nonatomic, assign) CGFloat scanLineY;
@property (nonatomic, assign) CGFloat scanLineDirection;
@property (nonatomic, retain) NSTimer *timer;
@end

@implementation ScannerOverlayView

- (instancetype)initWithFrame:(NSRect)frameRect {
    self = [super initWithFrame:frameRect];
    if (self) {
        self.wantsLayer = YES;
        self.progress = 0.0;
        self.statusText = @"Align camera with the QR code...";
        self.scanLineY = 60.0;
        self.scanLineDirection = 1.0;
        
        __weak typeof(self) weakSelf = self;
        self.timer = [NSTimer scheduledTimerWithTimeInterval:0.03 repeats:YES block:^(NSTimer * _Nonnull timer) {
            [weakSelf updateScanLine];
        }];
    }
    return self;
}

- (void)dealloc {
    [_timer invalidate];
}

- (void)updateScanLine {
    CGFloat height = self.bounds.size.height;
    self.scanLineY += self.scanLineDirection * 3.5;
    if (self.scanLineY >= height - 60 || self.scanLineY <= 60) {
        self.scanLineDirection *= -1;
    }
    self.needsDisplay = YES;
}

- (void)drawRect:(NSRect)dirtyRect {
    [super drawRect:dirtyRect];
    
    CGContextRef ctx = [[NSGraphicsContext currentContext] CGContext];
    if (!ctx) return;
    
    CGFloat width = self.bounds.size.width;
    CGFloat height = self.bounds.size.height;
    
    // 1. Draw Scanner Corners (Reticle) in the center
    CGFloat reticleSize = 280.0;
    CGRect reticleRect = CGRectMake((width - reticleSize) / 2.0, (height - reticleSize) / 2.0, reticleSize, reticleSize);
    
    CGContextSetStrokeColorWithColor(ctx, [[NSColor systemGreenColor] colorWithAlphaComponent:0.6].CGColor);
    CGContextSetLineWidth(ctx, 4.0);
    
    CGFloat cornerLen = 30.0;
    
    // Top-left corner
    CGContextMoveToPoint(ctx, reticleRect.origin.x, CGRectGetMaxY(reticleRect) - cornerLen);
    CGContextAddLineToPoint(ctx, reticleRect.origin.x, CGRectGetMaxY(reticleRect));
    CGContextAddLineToPoint(ctx, reticleRect.origin.x + cornerLen, CGRectGetMaxY(reticleRect));
    
    // Top-right corner
    CGContextMoveToPoint(ctx, CGRectGetMaxX(reticleRect) - cornerLen, CGRectGetMaxY(reticleRect));
    CGContextAddLineToPoint(ctx, CGRectGetMaxX(reticleRect), CGRectGetMaxY(reticleRect));
    CGContextAddLineToPoint(ctx, CGRectGetMaxX(reticleRect), CGRectGetMaxY(reticleRect) - cornerLen);
    
    // Bottom-right corner
    CGContextMoveToPoint(ctx, CGRectGetMaxX(reticleRect), reticleRect.origin.y + cornerLen);
    CGContextAddLineToPoint(ctx, CGRectGetMaxX(reticleRect), reticleRect.origin.y);
    CGContextAddLineToPoint(ctx, CGRectGetMaxX(reticleRect) - cornerLen, reticleRect.origin.y);
    
    // Bottom-left corner
    CGContextMoveToPoint(ctx, reticleRect.origin.x + cornerLen, reticleRect.origin.y);
    CGContextAddLineToPoint(ctx, reticleRect.origin.x, reticleRect.origin.y);
    CGContextAddLineToPoint(ctx, reticleRect.origin.x, reticleRect.origin.y + cornerLen);
    
    CGContextStrokePath(ctx);
    
    // 2. Draw Moving Scan Line
    CGContextSetStrokeColorWithColor(ctx, [[NSColor systemGreenColor] colorWithAlphaComponent:0.8].CGColor);
    CGContextSetLineWidth(ctx, 2.0);
    CGContextMoveToPoint(ctx, reticleRect.origin.x + 10, self.scanLineY);
    CGContextAddLineToPoint(ctx, CGRectGetMaxX(reticleRect) - 10, self.scanLineY);
    CGContextStrokePath(ctx);
    
    // Draw linear gradient glow behind the scan line
    const CGFloat colors[] = {
        0.0, 1.0, 0.0, 0.25,
        0.0, 1.0, 0.0, 0.0
    };
    CGColorSpaceRef colorSpace = CGColorSpaceCreateDeviceRGB();
    CGGradientRef gradient = CGGradientCreateWithColorComponents(colorSpace, colors, NULL, 2);
    if (gradient) {
        CGContextDrawLinearGradient(ctx,
                                    gradient,
                                    CGPointMake(width / 2.0, self.scanLineY),
                                    CGPointMake(width / 2.0, self.scanLineY - self.scanLineDirection * 18.0),
                                    0);
        CGGradientRelease(gradient);
    }
    CGColorSpaceRelease(colorSpace);
    
    // 3. Draw Bottom UI Card (Glassmorphism overlay)
    CGRect uiBoxRect = CGRectMake(40, 30, width - 80, 80);
    NSBezierPath *path = [NSBezierPath bezierPathWithRoundedRect:uiBoxRect xRadius:16 yRadius:16];
    [[[NSColor blackColor] colorWithAlphaComponent:0.65] set];
    [path fill];
    
    [[[NSColor whiteColor] colorWithAlphaComponent:0.15] set];
    [path setLineWidth:1.0];
    [path stroke];
    
    // 4. Draw Status Text
    NSMutableParagraphStyle *textStyle = [[NSMutableParagraphStyle alloc] init];
    textStyle.alignment = NSTextAlignmentCenter;
    NSFont *textFont = [NSFont systemFontOfSize:13 weight:NSFontWeightSemibold];
    NSDictionary *textAttrs = @{
        NSFontAttributeName: textFont,
        NSForegroundColorAttributeName: [NSColor whiteColor],
        NSParagraphStyleAttributeName: textStyle
    };
    
    CGRect textRect = CGRectMake(uiBoxRect.origin.x + 10, CGRectGetMaxY(uiBoxRect) - 32, uiBoxRect.size.width - 20, 22);
    [self.statusText drawInRect:textRect withAttributes:textAttrs];
    
    // 5. Draw Progress Bar Background
    CGRect pBarBgRect = CGRectMake(uiBoxRect.origin.x + 30, uiBoxRect.origin.y + 18, uiBoxRect.size.width - 60, 8);
    NSBezierPath *bgPath = [NSBezierPath bezierPathWithRoundedRect:pBarBgRect xRadius:4 yRadius:4];
    [[[NSColor whiteColor] colorWithAlphaComponent:0.12] set];
    [bgPath fill];
    
    // Draw Progress Bar Fill
    if (self.progress > 0) {
        CGRect pBarFillRect = CGRectMake(pBarBgRect.origin.x, pBarBgRect.origin.y, pBarBgRect.size.width * self.progress, 8);
        NSBezierPath *fillPath = [NSBezierPath bezierPathWithRoundedRect:pBarFillRect xRadius:4 yRadius:4];
        [[NSColor systemGreenColor] set];
        [fillPath fill];
    }
}
@end

@interface AppDelegate : NSObject <NSApplicationDelegate, NSWindowDelegate, AVCaptureVideoDataOutputSampleBufferDelegate>
@property (nonatomic, retain) NSWindow *window;
@property (nonatomic, retain) AVCaptureSession *captureSession;
@property (nonatomic, retain) AVCaptureVideoPreviewLayer *previewLayer;
@property (nonatomic, retain) ScannerOverlayView *overlayView;
@property (nonatomic, retain) AVCaptureVideoDataOutput *videoDataOutput;
@property (nonatomic, retain) dispatch_queue_t videoQueue;

// Chunk tracking
@property (nonatomic, retain) NSMutableDictionary *receivedChunks;
@property (nonatomic, assign) NSInteger totalChunks;
@end

@implementation AppDelegate

- (void)applicationDidFinishLaunching:(NSNotification *)notification {
    self.receivedChunks = [NSMutableDictionary dictionary];
    self.totalChunks = -1;
    
    CGFloat windowWidth = 640.0;
    CGFloat windowHeight = 480.0;
    
    NSRect screenRect = [[NSScreen mainScreen] frame];
    NSRect windowRect = NSMakeRect((screenRect.size.width - windowWidth) / 2.0,
                                   (screenRect.size.height - windowHeight) / 2.0,
                                   windowWidth,
                                   windowHeight);
    
    self.window = [[NSWindow alloc] initWithContentRect:windowRect
                                              styleMask:(NSWindowStyleMaskTitled | NSWindowStyleMaskClosable | NSWindowStyleMaskFullSizeContentView)
                                                backing:NSBackingStoreBuffered
                                                  defer:NO];
    self.window.title = @"Offline Text Scanner";
    self.window.titlebarAppearsTransparent = YES;
    self.window.titleVisibility = NSWindowTitleHidden;
    self.window.level = NSFloatingWindowLevel;
    self.window.delegate = self;
    self.window.movableByWindowBackground = YES;
    
    NSView *contentView = [[NSView alloc] initWithFrame:self.window.contentView.bounds];
    contentView.wantsLayer = YES;
    contentView.layer.backgroundColor = [NSColor blackColor].CGColor;
    self.window.contentView = contentView;
    
    [self setupCamera];
    
    self.overlayView = [[ScannerOverlayView alloc] initWithFrame:contentView.bounds];
    self.overlayView.autoresizingMask = (NSViewWidthSizable | NSViewHeightSizable);
    [contentView addSubview:self.overlayView];
    
    [self.window makeKeyAndOrderFront:nil];
    [NSApp activateIgnoringOtherApps:YES];
}

- (void)setupCamera {
    self.captureSession = [[AVCaptureSession alloc] init];
    self.captureSession.sessionPreset = AVCaptureSessionPreset640x480;
    
    AVCaptureDevice *videoDevice = [AVCaptureDevice defaultDeviceWithMediaType:AVMediaTypeVideo];
    if (!videoDevice) {
        [self showErrorAndExit:@"No camera device found."];
        return;
    }
    
    NSError *error = nil;
    AVCaptureDeviceInput *videoInput = [AVCaptureDeviceInput deviceInputWithDevice:videoDevice error:&error];
    if (error || !videoInput) {
        [self showErrorAndExit:[NSString stringWithFormat:@"Cannot init camera: %@", error.localizedDescription]];
        return;
    }
    
    if ([self.captureSession canAddInput:videoInput]) {
        [self.captureSession addInput:videoInput];
    } else {
        [self showErrorAndExit:@"Cannot add video camera input."];
        return;
    }
    
    self.videoDataOutput = [[AVCaptureVideoDataOutput alloc] init];
    self.videoDataOutput.alwaysDiscardsLateVideoFrames = YES;
    self.videoQueue = dispatch_queue_create("videoQueue", DISPATCH_QUEUE_SERIAL);
    [self.videoDataOutput setSampleBufferDelegate:self queue:self.videoQueue];
    
    if ([self.captureSession canAddOutput:self.videoDataOutput]) {
        [self.captureSession addOutput:self.videoDataOutput];
    }
    
    self.previewLayer = [AVCaptureVideoPreviewLayer layerWithSession:self.captureSession];
    self.previewLayer.videoGravity = AVLayerVideoGravityResizeAspectFill;
    self.previewLayer.frame = self.window.contentView.bounds;
    [self.window.contentView.layer addSublayer:self.previewLayer];
    
    AVAuthorizationStatus status = [AVCaptureDevice authorizationStatusForMediaType:AVMediaTypeVideo];
    if (status == AVAuthorizationStatusAuthorized) {
        [self.captureSession startRunning];
    } else if (status == AVAuthorizationStatusNotDetermined) {
        __weak typeof(self) weakSelf = self;
        [AVCaptureDevice requestAccessForMediaType:AVMediaTypeVideo completionHandler:^(BOOL granted) {
            dispatch_async(dispatch_get_main_queue(), ^{
                if (granted) {
                    [weakSelf.captureSession startRunning];
                } else {
                    [weakSelf showErrorAndExit:@"Camera access permission denied."];
                }
            });
        }];
    } else {
        [self showErrorAndExit:@"Camera access is restricted or denied."];
    }
}

- (void)captureOutput:(AVCaptureOutput *)output didOutputSampleBuffer:(CMSampleBufferRef)sampleBuffer fromConnection:(AVCaptureConnection *)connection {
    CVImageBufferRef imageBuffer = CMSampleBufferGetImageBuffer(sampleBuffer);
    if (!imageBuffer) return;
    
    CIImage *ciImage = [CIImage imageWithCVImageBuffer:imageBuffer];
    
    NSDictionary *options = @{CIDetectorAccuracy: CIDetectorAccuracyHigh};
    CIDetector *detector = [CIDetector detectorOfType:CIDetectorTypeQRCode context:nil options:options];
    
    NSArray *features = [detector featuresInImage:ciImage];
    for (CIFeature *feature in features) {
        if ([feature isKindOfClass:[CIQRCodeFeature class]]) {
            CIQRCodeFeature *qrFeature = (CIQRCodeFeature *)feature;
            NSString *message = qrFeature.messageString;
            if (message) {
                [self processQRCodeMessage:message];
            }
        }
    }
}

- (void)processQRCodeMessage:(NSString *)message {
    NSString *pattern = @"^(\\d+)/(\\d+):(.*)$";
    NSError *error = nil;
    NSRegularExpression *regex = [NSRegularExpression regularExpressionWithPattern:pattern
                                                                           options:NSRegularExpressionDotMatchesLineSeparators
                                                                             error:&error];
    if (error || !regex) return;
    
    NSTextCheckingResult *match = [regex firstMatchInString:message options:0 range:NSMakeRange(0, message.length)];
    if (!match) return;
    
    NSString *indexStr = [message substringWithRange:[match rangeAtIndex:1]];
    NSString *totalStr = [message substringWithRange:[match rangeAtIndex:2]];
    NSString *payload = [message substringWithRange:[match rangeAtIndex:3]];
    
    NSInteger index = [indexStr integerValue];
    NSInteger total = [totalStr integerValue];
    
    dispatch_async(dispatch_get_main_queue(), ^{
        [self handleChunkReceivedWithIndex:index total:total payload:payload];
    });
}

- (void)handleChunkReceivedWithIndex:(NSInteger)index total:(NSInteger)total payload:(NSString *)payload {
    if (self.totalChunks != -1 && self.totalChunks != total) {
        [self.receivedChunks removeAllObjects];
    }
    self.totalChunks = total;
    
    NSNumber *key = @(index);
    if (!self.receivedChunks[key]) {
        self.receivedChunks[key] = payload;
        
        // Play audio feedback for successful chunk detection
        [[NSSound soundNamed:@"Tink"] play];
        
        CGFloat progressRatio = (CGFloat)self.receivedChunks.count / (CGFloat)total;
        self.overlayView.progress = progressRatio;
        self.overlayView.statusText = [NSString stringWithFormat:@"Received chunk %lu of %ld...", (unsigned long)self.receivedChunks.count, (long)total];
        self.overlayView.needsDisplay = YES;
        
        if (self.receivedChunks.count == total) {
            [self finishScanning];
        }
    }
}

- (void)finishScanning {
    [self.captureSession stopRunning];
    
    NSMutableString *fullText = [NSMutableString string];
    for (NSInteger i = 1; i <= self.totalChunks; i++) {
        NSString *chunk = self.receivedChunks[@(i)];
        if (chunk) {
            [fullText appendString:chunk];
        }
    }
    
    NSPasteboard *pc = [NSPasteboard generalPasteboard];
    [pc clearContents];
    [pc setString:fullText forType:NSPasteboardTypeString];
    
    [[NSSound soundNamed:@"Glass"] play];
    
    self.overlayView.statusText = @"Complete! Copied to clipboard.";
    self.overlayView.progress = 1.0;
    self.overlayView.needsDisplay = YES;
    
    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.6 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
        [NSApp terminate:nil];
    });
}

- (void)showErrorAndExit:(NSString *)message {
    dispatch_async(dispatch_get_main_queue(), ^{
        fprintf(stderr, "Error: %s\n", [message UTF8String]);
        NSAlert *alert = [[NSAlert alloc] init];
        alert.messageText = @"QR Scanner Error";
        alert.informativeText = message;
        alert.alertStyle = NSAlertStyleCritical;
        [alert addButtonWithTitle:@"OK"];
        [alert runModal];
        [NSApp terminate:nil];
    });
}

- (BOOL)windowShouldClose:(NSWindow *)sender {
    [NSApp terminate:nil];
    return YES;
}
@end

int main(int argc, const char * argv[]) {
    @autoreleasepool {
        NSApplication *app = [NSApplication sharedApplication];
        AppDelegate *delegate = [[AppDelegate alloc] init];
        app.delegate = delegate;
        [app setActivationPolicy:NSApplicationActivationPolicyAccessory];
        [app run];
    }
    return 0;
}

File: qr-scanner.swift
import Cocoa
import AVFoundation
import CoreImage
import CoreMedia
import CoreVideo

class ScannerOverlayView: NSView {
    var progress: CGFloat = 0.0
    var statusText: String = "Align camera with the QR code..."
    var scanLineY: CGFloat = 0.0
    var scanLineDirection: CGFloat = 1.0
    var timer: Timer?
    
    override init(frame frameRect: NSRect) {
        super.init(frame: frameRect)
        wantsLayer = true
        timer = Timer.scheduledTimer(withTimeInterval: 0.03, repeats: true) { [weak self] _ in
            self?.updateScanLine()
        }
    }
    
    required init?(coder: NSCoder) {
        super.init(coder: coder)
    }
    
    deinit {
        timer?.invalidate()
    }
    
    func updateScanLine() {
        let height = bounds.height
        // Animate scan line up and down
        scanLineY += scanLineDirection * 3.5
        if scanLineY >= height - 60 || scanLineY <= 60 {
            scanLineDirection *= -1
        }
        needsDisplay = true
    }
    
    override func draw(_ dirtyRect: NSRect) {
        super.draw(dirtyRect)
        
        guard let ctx = NSGraphicsContext.current?.cgContext else { return }
        
        let width = bounds.width
        let height = bounds.height
        
        // 1. Draw Scanner Corners (Reticle) in the center
        let reticleSize: CGFloat = 280
        let reticleRect = CGRect(
            x: (width - reticleSize) / 2,
            y: (height - reticleSize) / 2,
            width: reticleSize,
            height: reticleSize
        )
        
        ctx.setStrokeColor(NSColor.systemGreen.withAlphaComponent(0.6).cgColor)
        ctx.setLineWidth(4.0)
        
        let cornerLen: CGFloat = 30
        
        // Top-left corner
        ctx.move(to: CGPoint(x: reticleRect.minX, y: reticleRect.maxY - cornerLen))
        ctx.addLine(to: CGPoint(x: reticleRect.minX, y: reticleRect.maxY))
        ctx.addLine(to: CGPoint(x: reticleRect.minX + cornerLen, y: reticleRect.maxY))
        
        // Top-right corner
        ctx.move(to: CGPoint(x: reticleRect.maxX - cornerLen, y: reticleRect.maxY))
        ctx.addLine(to: CGPoint(x: reticleRect.maxX, y: reticleRect.maxY))
        ctx.addLine(to: CGPoint(x: reticleRect.maxX, y: reticleRect.maxY - cornerLen))
        
        // Bottom-right corner
        ctx.move(to: CGPoint(x: reticleRect.maxX, y: reticleRect.minY + cornerLen))
        ctx.addLine(to: CGPoint(x: reticleRect.maxX, y: reticleRect.minY))
        ctx.addLine(to: CGPoint(x: reticleRect.maxX - cornerLen, y: reticleRect.minY))
        
        // Bottom-left corner
        ctx.move(to: CGPoint(x: reticleRect.minX + cornerLen, y: reticleRect.minY))
        ctx.addLine(to: CGPoint(x: reticleRect.minX, y: reticleRect.minY))
        ctx.addLine(to: CGPoint(x: reticleRect.minX, y: reticleRect.minY + cornerLen))
        
        ctx.strokePath()
        
        // 2. Draw Moving Scan Line
        ctx.setStrokeColor(NSColor.systemGreen.withAlphaComponent(0.8).cgColor)
        ctx.setLineWidth(2.0)
        ctx.move(to: CGPoint(x: reticleRect.minX + 10, y: scanLineY))
        ctx.addLine(to: CGPoint(x: reticleRect.maxX - 10, y: scanLineY))
        ctx.strokePath()
        
        // Draw linear gradient glow behind the scan line
        let colors = [
            NSColor.systemGreen.withAlphaComponent(0.25).cgColor,
            NSColor.clear.cgColor
        ]
        let colorSpace = CGColorSpaceCreateDeviceRGB()
        if let gradient = CGGradient(colorsSpace: colorSpace, colors: colors as CFArray, locations: [0.0, 1.0]) {
            ctx.drawLinearGradient(
                gradient,
                start: CGPoint(x: width / 2, y: scanLineY),
                end: CGPoint(x: width / 2, y: scanLineY - scanLineDirection * 18),
                options: []
            )
        }
        
        // 3. Draw Bottom UI Card (Glassmorphism overlay)
        let uiBoxRect = CGRect(x: 40, y: 30, width: width - 80, height: 80)
        let path = NSBezierPath(roundedRect: uiBoxRect, xRadius: 16, yRadius: 16)
        NSColor.black.withAlphaComponent(0.65).set()
        path.fill()
        
        NSColor.white.withAlphaComponent(0.15).set()
        path.lineWidth = 1.0
        path.stroke()
        
        // 4. Draw Status Text
        let textStyle = NSMutableParagraphStyle()
        textStyle.alignment = .center
        let textFont = NSFont.systemFont(ofSize: 13, weight: .semibold)
        let textAttrs: [NSAttributedString.Key: Any] = [
            .font: textFont,
            .foregroundColor: NSColor.white,
            .paragraphStyle: textStyle
        ]
        
        let textRect = CGRect(x: uiBoxRect.minX + 10, y: uiBoxRect.maxY - 32, width: uiBoxRect.width - 20, height: 22)
        statusText.draw(in: textRect, withAttributes: textAttrs)
        
        // 5. Draw Progress Bar background
        let pBarBgRect = CGRect(x: uiBoxRect.minX + 30, y: uiBoxRect.minY + 18, width: uiBoxRect.width - 60, height: 8)
        let bgPath = NSBezierPath(roundedRect: pBarBgRect, xRadius: 4, yRadius: 4)
        NSColor.white.withAlphaComponent(0.12).set()
        bgPath.fill()
        
        // Draw Progress Bar Fill
        if progress > 0 {
            let pBarFillRect = CGRect(x: pBarBgRect.minX, y: pBarBgRect.minY, width: pBarBgRect.width * progress, height: 8)
            let fillPath = NSBezierPath(roundedRect: pBarFillRect, xRadius: 4, yRadius: 4)
            NSColor.systemGreen.set()
            fillPath.fill()
        }
    }
}

class AppDelegate: NSObject, NSApplicationDelegate, NSWindowDelegate, AVCaptureVideoDataOutputSampleBufferDelegate {
    var window: NSWindow!
    var captureSession: AVCaptureSession!
    var previewLayer: AVCaptureVideoPreviewLayer!
    var overlayView: ScannerOverlayView!
    var videoDataOutput: AVCaptureVideoDataOutput!
    var videoQueue: DispatchQueue!
    
    // Chunk tracking
    var receivedChunks: [Int: String] = [:]
    var totalChunks: Int? = nil
    
    func applicationDidFinishLaunching(_ notification: Notification) {
        let windowWidth: CGFloat = 640
        let windowHeight: CGFloat = 480
        
        let screenRect = NSScreen.main?.frame ?? CGRect(x: 0, y: 0, width: 1024, height: 768)
        let windowRect = CGRect(
            x: (screenRect.width - windowWidth) / 2,
            y: (screenRect.height - windowHeight) / 2,
            width: windowWidth,
            height: windowHeight
        )
        
        window = NSWindow(
            contentRect: windowRect,
            styleMask: [.titled, .closable, .fullSizeContentView],
            backing: .buffered,
            defer: false
        )
        
        window.title = "Offline Text Scanner"
        window.titlebarAppearsTransparent = true
        window.titleVisibility = .hidden
        window.level = .floating // Floating on top of other windows
        window.delegate = self
        window.isMovableByWindowBackground = true
        
        let contentView = NSView(frame: window.contentView!.bounds)
        contentView.wantsLayer = true
        contentView.layer?.backgroundColor = NSColor.black.cgColor
        window.contentView = contentView
        
        setupCamera()
        
        overlayView = ScannerOverlayView(frame: contentView.bounds)
        overlayView.autoresizingMask = [.width, .height]
        contentView.addSubview(overlayView)
        
        window.makeKeyAndOrderFront(nil)
        NSApp.activate(ignoringOtherApps: true)
    }
    
    func setupCamera() {
        captureSession = AVCaptureSession()
        captureSession.sessionPreset = .v1_640x480
        
        guard let videoDevice = AVCaptureDevice.default(for: .video) else {
            showErrorAndExit(message: "No camera device found.")
            return
        }
        
        do {
            let videoInput = try AVCaptureDeviceInput(device: videoDevice)
            if captureSession.canAddInput(videoInput) {
                captureSession.addInput(videoInput)
            } else {
                showErrorAndExit(message: "Cannot add video camera input.")
                return
            }
        } catch {
            showErrorAndExit(message: "Camera initialization failed: \(error.localizedDescription)")
            return
        }
        
        videoDataOutput = AVCaptureVideoDataOutput()
        videoDataOutput.alwaysDiscardsLateVideoFrames = true
        videoQueue = DispatchQueue(label: "videoQueue", qos: .userInteractive)
        videoDataOutput.setSampleBufferDelegate(self, queue: videoQueue)
        
        if captureSession.canAddOutput(videoDataOutput) {
            captureSession.addOutput(videoDataOutput)
        }
        
        previewLayer = AVCaptureVideoPreviewLayer(session: captureSession)
        previewLayer.videoGravity = .resizeAspectFill
        previewLayer.frame = window.contentView!.bounds
        window.contentView!.layer?.addSublayer(previewLayer)
        
        // Request/check authorization status
        let status = AVCaptureDevice.authorizationStatus(for: .video)
        if status == .authorized {
            captureSession.startRunning()
        } else if status == .notDetermined {
            AVCaptureDevice.requestAccess(for: .video) { [weak self] granted in
                DispatchQueue.main.async {
                    if granted {
                        self?.captureSession.startRunning()
                    } else {
                        self?.showErrorAndExit(message: "Camera access permission denied.")
                    }
                }
            }
        } else {
            showErrorAndExit(message: "Camera access is restricted or denied.")
        }
    }
    
    func captureOutput(_ output: AVCaptureOutput, didOutput sampleBuffer: CMSampleBuffer, from connection: AVCaptureConnection) {
        guard let imageBuffer = CMSampleBufferGetImageBuffer(sampleBuffer) else { return }
        
        let ciImage = CIImage(cvImageBuffer: imageBuffer)
        
        // Use high accuracy detector for reliable scanner reading
        let options = [CIDetectorAccuracy: CIDetectorAccuracyHigh]
        guard let detector = CIDetector(ofType: CIDetectorTypeQRCode, context: nil, options: options) else { return }
        
        let features = detector.features(in: ciImage)
        
        for feature in features {
            if let qrFeature = feature as? CIQRCodeFeature, let message = qrFeature.messageString {
                processQRCodeMessage(message)
            }
        }
    }
    
    func processQRCodeMessage(_ message: String) {
        // Regex format matching "index/total:payload"
        let pattern = "^(\\d+)/(\\d+):(.*)$"
        guard let regex = try? NSRegularExpression(pattern: pattern, options: [.dotMatchesLineSeparators]) else { return }
        
        let nsString = message as NSString
        let results = regex.matches(in: message, options: [], range: NSRange(location: 0, length: nsString.length))
        
        guard let match = results.first else { return }
        
        let indexStr = nsString.substring(with: match.range(at: 1))
        let totalStr = nsString.substring(with: match.range(at: 2))
        let payload = nsString.substring(with: match.range(at: 3))
        
        guard let index = Int(indexStr), let total = Int(totalStr) else { return }
        
        DispatchQueue.main.async { [weak self] in
            self?.handleChunkReceived(index: index, total: total, payload: payload)
        }
    }
    
    func handleChunkReceived(index: Int, total: Int, payload: String) {
        if let currentTotal = totalChunks {
            if currentTotal != total {
                // If total chunks count differs, reset state
                receivedChunks.removeAll()
            }
        }
        totalChunks = total
        
        if receivedChunks[index] == nil {
            receivedChunks[index] = payload
            
            // Play audio feedback for successful frame detection
            NSSound(named: "Tink")?.play()
            
            let progressRatio = CGFloat(receivedChunks.count) / CGFloat(total)
            overlayView.progress = progressRatio
            overlayView.statusText = "Received chunk \(receivedChunks.count) of \(total)..."
            overlayView.needsDisplay = true
            
            if receivedChunks.count == total {
                finishScanning()
            }
        }
    }
    
    func finishScanning() {
        captureSession.stopRunning()
        
        guard let total = totalChunks else { return }
        var fullText = ""
        for i in 1...total {
            if let chunk = receivedChunks[i] {
                fullText += chunk
            }
        }
        
        // Write the fully assembled text to pasteboard
        NSPasteboard.general.clearContents()
        NSPasteboard.general.setString(fullText, forType: .string)
        
        // Play success sound
        NSSound(named: "Glass")?.play()
        
        overlayView.statusText = "Complete! Copied to clipboard."
        overlayView.progress = 1.0
        overlayView.needsDisplay = true
        
        // Exit app after a short delay
        DispatchQueue.main.asyncAfter(deadline: .now() + 0.6) {
            NSApp.terminate(nil)
        }
    }
    
    func showErrorAndExit(message: String) {
        DispatchQueue.main.async {
            fputs("Error: \(message)\n", stderr)
            let alert = NSAlert()
            alert.messageText = "QR Scanner Error"
            alert.informativeText = message
            alert.alertStyle = .critical
            alert.addButton(withTitle: "OK")
            alert.runModal()
            NSApp.terminate(nil)
        }
    }
    
    func windowShouldClose(_ sender: NSWindow) -> Bool {
        NSApp.terminate(nil)
        return true
    }
}

let app = NSApplication.shared
let delegate = AppDelegate()
app.delegate = delegate
app.setActivationPolicy(.accessory)
app.run()
